### ISSUE TITLE
[java] SwitchDensity false positive with pattern matching

### ISSUE THREAD

**Issue by cowwoc** [pmd/pmd#5030]
**Affects PMD Version:** 7.2.0

**Rule:** SwitchDensity

Please provide the rule name and a link to the rule documentation:
https://pmd.github.io/pmd/pmd_rules_java_design.html#switchdensity

**Description:**

Given the following code, there seems to be ~4 statements per switch label but the rule seems to think the density is greater than 10.

**Code Sample demonstrating the issue:**

```java
private void handleException(RequestScope scope, Throwable exception)
{
	ITransaction sentryTransaction = scope.getSentryTransaction();
	if (exception.getClass() != RequestFailedException.class)
	{
		// No need to report RequestFailedException as it's just used to unwind the stack
		sentryTransaction.setThrowable(exception);
	}
	switch (exception)
	{
		case JsonProcessingException e ->
		{
			log.debug("", e);
			responseToJacksonException(scope, exception);
			sentryTransaction.setStatus(SpanStatus.fromHttpStatusCode(scope.getResponse().getStatus()));
		}
		case RequestFailedException _ ->
		{
			// The exception is just used to unroll the stack after sending a response. No need to log it.
			sentryTransaction.setStatus(SpanStatus.fromHttpStatusCode(scope.getResponse().getStatus()));
		}
		case DataAccessException e ->
		{
			log.error("", e);
			scope.getRequests().sendText(INTERNAL_SERVER_ERROR_500, Exceptions.getDeepestMessage(e),
				HttpFields.EMPTY);
			sentryTransaction.setStatus(SpanStatus.fromHttpStatusCode(scope.getResponse().getStatus()));
		}
		case RuntimeException e ->
		{
			log.error("", e);
			scope.getRequests().sendText(BAD_REQUEST_400, Exceptions.getDeepestMessage(e), HttpFields.EMPTY);
			sentryTransaction.setStatus(SpanStatus.fromHttpStatusCode(scope.getResponse().getStatus()));
		}
		case Error e ->
		{
			scope.getRequests().sendText(INTERNAL_SERVER_ERROR_500, Exceptions.getDeepestMessage(e),
				HttpFields.EMPTY);
			sentryTransaction.setStatus(SpanStatus.fromHttpStatusCode(scope.getResponse().getStatus()));
			throw e;
		}
		case Throwable t ->
		{
			scope.getRequests().sendText(INTERNAL_SERVER_ERROR_500, Exceptions.getDeepestMessage(t),
				HttpFields.EMPTY);
			sentryTransaction.setStatus(SpanStatus.fromHttpStatusCode(scope.getResponse().getStatus()));
			throw WrappedCheckedException.wrap(t);
		}
	}
}
```

**Running PMD through:** maven-pmd-plugin 3.22.0

**Comment by cowwoc** [comment]
There is definitely a problem here. Even this configuration triggers a violation:

```
	<rule ref="category/java/design.xml/SwitchDensity">
		<properties>
			<property name="minimum" value="100"/>
		</properties>
	</rule>
```

**Comment by Gold856** [comment]
PMD reports Infinity for the switch density; it detects zero labels. The parser probably doesn't recognize `case Type t ->` as a label.

**Comment by oowekyala** [comment]
Yeah the problem is that the `getExprList` is empty when the branch is a pattern:
https://github.com/pmd/pmd/blob/f30b535f61737f18f5ded94f3261f6314d26814b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/SwitchDensityRule.java#L58

This could be changed to be for instance
```java
label.isDefault() || label.isPatternLabel() ? 1 : label.getExprList().count()
```
Evidently the documentation [here](https://github.com/pmd/pmd/blob/f30b535f61737f18f5ded94f3261f6314d26814b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTSwitchLabel.java#L53) does not match the implementation as patterns are not pushed in the tree as PatternExpressions. [This line](https://github.com/pmd/pmd/blob/f30b535f61737f18f5ded94f3261f6314d26814b/pmd-java/etc/grammar/Java.jjt#L2653) should have read `#PatternExpression` instead of `#void`. I think it is probably best not to change the grammar at this point, but to add some methods to `ASTSwitchLabel` to handle the case where there is a pattern, eg `isPatternLabel()`, `getPattern()`, `getGuard()`.

**Comment by jsotuyod** [comment]
Yeah, we should not change the grammar for compatibility purposes, but we will need to track this for PMD 8 (early next year).
